### PR TITLE
MNT: migrate to GitHub actions and pyproject.toml [LCLSPC-603]

### PR DESCRIPTION
**Pull request by tangkong**
## Description
* Upgrading by way of https://github.com/pcdshub/pcds-python-migration-tools
* elog no longer uses Travis CI and has migrated to GitHub Actions for continuous integration, testing, and documentation deployment.
* elog has been migrated to use setuptools-scm, replacing versioneer, as its version-string management tool of choice.
* elog has been migrated to use the modern ``pyproject.toml``, replacing ``setup.py`` and related files.
* Older language features and syntax found in the repository have been updated to Python 3.9+ standards by way of ``pyupgrade``.
* `pyupgrade` was used to update annotations.
* `pycln` was used to clean up imports

## Motivation and Context
Migrating all python packages to GHA

## How Has This Been Tested?
Running CI here

## Where Has This Been Documented?
This PR and Jira [LCLSPC-603](https://jira.slac.stanford.edu/browse/LCLSPC-603)

### CODE REVIEW

**Comment by tangkong** [comment]
Elog specific intricacies: 
* no docs -> test fails automatically
* missing `gssapi/gssapi.h` requires `libkrb5-dev` installation, luckily `pip-system-packages` was added recently ([log traceback](https://github.com/pcdshub/elog/actions/runs/4255878603/jobs/7404053055#step:10:87))
* Dependencies from an external url need to be listed in a slightly different way than before (`lcls-krtc @ git+https://github.com/slaclab/krtc@master
`), but can still live in `requirements.txt`
* `[project.scripts]` replaces the `scripts` argument in `setup.py`, looks a bit like an entrypoint

File: pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+dynamic = [ "version", "readme", "dependencies", "optional-dependencies", "optional-dependencies",]
+keywords = []
+name = "elog"
+requires-python = ">=3.9"
+
+[[project.authors]]
+name = "SLAC National Accelerator Laboratory"
+
+[project.scripts]
+LogBookPost = "elog.scripts.LogBookPost:main"
```
Comment (klauer) [inline pyproject.toml:17]: I assume you tried this out and it works? 👍 if so

File: pyproject.toml [hunk 2]
```diff
@@ -0,0 +1,41 @@
+[build-system]
+build-backend = "setuptools.build_meta"
+requires = [ "setuptools>=45", "setuptools_scm[toml]>=6.2",]
+
+[project]
+classifiers = [ "Development Status :: 2 - Pre-Alpha", "Natural Language :: English", "Programming Language :: Python :: 3",]
```
Comment (klauer) [inline pyproject.toml:6]: Preferred classifier here @ZLLentz? I'd guess production

File: conda-recipe/meta.yaml [hunk 3]
```diff
@@ -1,29 +1,39 @@
...
 
 source:
   path: ..
 
 build:
-  number: 1
+  number: 0
   noarch: python
+  script: {{ PYTHON }} -m pip install . -vv
+
```
Comment (klauer) [inline conda-recipe/meta.yaml:16]: Nitpicky - migration tools add a bunch of whitespace here. Probably good to clean it up.

File: elog/scripts/LogBookPost.py [hunk 4]
```diff
@@ -74,16 +75,20 @@
...
+                                 user=args.user,
+                                 pw=args.password,
+                                 base_url=args.webserviceurl)
         params.update({"experiment": True, "facility": False})
-        elog.post(**params)
+        elogger.post(**params)
+
+
+if __name__ == '__main__':
+    main()
```
Comment (klauer) [inline elog/scripts/LogBookPost.py:94]: (Picking a random line to comment)

Probably should add `elog.tests.__init__` to make it a proper submodule

**Review by klauer — APPROVED** [review]
Tests passing (aside from missing docs) - this all looks good to me 👍

File: pyproject.toml (hunk 2 above)
Comment (ZLLentz) [reply pyproject.toml:6]: Development Status :: 5 - Production/Stable is probably the most accurate

File: pyproject.toml (hunk 1 above)
Comment (ZLLentz) [reply pyproject.toml:17]: I didn't even know this had a script built-in for that

**Review by ZLLentz — APPROVED** [review]
Looks like a gha to me

File: pyproject.toml (hunk 1 above)
Comment (tangkong) [reply pyproject.toml:17]: Yep, locally at least 👍

File: requirements.txt [hunk 5]
```diff
@@ -1,3 +1,3 @@
 requests
-git+https://github.com/slaclab/krtc
+lcls-krtc @ git+https://github.com/slaclab/krtc@master
```
[the remark this answers is not in the document]
Comment (tangkong) [reply requirements.txt:2]: This isn't what I would have guessed either

File: MANIFEST.in [hunk 6]
```diff
@@ -1,2 +1,11 @@
 include versioneer.py
```
Comment (klauer) [inline MANIFEST.in:1]:
```suggestion
```

Comment (tangkong) [reply MANIFEST.in:1]: ah, right

**Comment by tangkong** [comment]
I've made the documentation test optional until we actually ... have docs.  Looks like gha, sounds like gha, will be merged as gha.

**Pull request merged**